internal/dynamic/plugin/config.go

```diff
@@ -10,11 +10,283 @@
+		}
+	case FieldKindMap:
+		switch fieldType {
+		case FieldTypeBool:
+			// TODO: This should be a BoolMapField, but we don't have one yet.
+			f = service.NewAnyMapField(c.Name)
+		case FieldTypeFloat:
+			f = service.NewFloatMapField(c.Name)
+		case FieldTypeInt:
+			f = service.NewIntMapField(c.Name)
+		case FieldTypeString:
+			f = service.NewStringMapField(c.Name)
+		case FieldTypeUnknown:
+			f = service.NewAnyMapField(c.Name)
+		default:
+			return nil, fmt.Errorf("unexpected plugin.FieldType: %#v", fieldType)
+		}
+	default:
+		return nil, fmt.Errorf("unexpected plugin.FieldKind: %#v", fieldKind)
+	}
+	if c.Default != nil {
+		f = f.Default(*c.Default)
+	}
+	if c.Advanced {
+		f = f.Advanced()
+	}
+	if c.Description != "" {
+		f = f.Description(c.Description)
+	}
+	return f, nil
+}
+
+// Validate checks that the field config is valid.
+func (c *FieldConfig) Validate() error {
+	if c.Name == "" {
+		return errors.New("field name is required")
+	}
+	if c.Type != nil {
+		if err := c.Type.Validate(); err != nil {
+			return err
+		}
+	}
+	if c.Kind != nil {
+		if err := c.Kind.Validate(); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// PluginType describes the type of plugin.
+type PluginType string
+
+// Validate checks that the plugin type is valid.
+func (p PluginType) Validate() error {
+	switch p {
+	case PluginTypeInput, PluginTypeProcessor, PluginTypeOutput:
+		return nil
+	}
+	return fmt.Errorf("invalid plugin type: %q", p)
+}
+
+const (
+	PluginTypeInput     PluginType = "input"
+	PluginTypeProcessor PluginType = "processor"
+	PluginTypeOutput    PluginType = "output"
+)
+
+// Config describes a dynamic plugin over gRPC.
 type Config struct {
+	Name        string `yaml:"name"`
+	Summary     string `yaml:"summary"`
+	Description string `yaml:"description"`
+	// The command to run for the plugin.
+	Cmd    []string      `yaml:"command"`
+	Type   PluginType    `yaml:"type"`
+	Fields []FieldConfig `yaml:"fields"`
+}
+
+// Validate checks that the config is valid.
+func (c *Config) Validate() error {
+	if c.Name == "" {
+		return errors.New("plugin name is required")
+	}
+	if len(c.Cmd) == 0 {
+		return errors.New("plugin command is required")
+	}
+	if err := c.Type.Validate(); err != nil {
+		return err
+	}
+	for _, field := range c.Fields {
+		if err := field.Validate(); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (c *Config) toSpec() (*service.ConfigSpec, error) {
+	spec := service.NewConfigSpec()
+	if c.Summary != "" {
+		spec = spec.Summary(c.Summary)
+	}
+	if c.Description != "" {
+		spec = spec.Description(c.Description)
+	}
+	for _, field := range c.Fields {
+		fieldSpec, err := field.toSpec()
+		if err != nil {
+			return nil, err
+		}
+		spec = spec.Field(fieldSpec)
+	}
+	return spec, nil
+}
+
+func DiscoverAndRegisterPlugins(fs fs.FS, env *service.Environment, paths []string) error {
+	paths, err := service.Globs(fs, paths...)
```

Comment:
Paths redeclared, it would be worth documenting that paths may contain glob patterns.

Reply:
my bad, it's over 5K LOC, I read it commit by commit

Reply:
No you're good! I just respond to every comment. I should keep m commit history cleaner, I generally am sloppy in this repo